architecture/2025-02-user-defined-resource-type-feature-spec.md

```diff
@@ -0,0 +1,1212 @@
+    }     
+
+  }
+}
+
+resource "applications_core_containers" "serviceContainer" {
+  name                = "${var.context.resource.name}-container"
+  environment         = var.context.environment.id
+  application         = var.context.application.id
+  container_image     = context.resource.properties.container.image
+  ...
+}
+```
+
+**User Experience 2 – Using Bicep template**
+
+The recipe will have a conditional on the ingress property to determine whether to deploy a gateway resource.
+
+**`webservice-recipe.bicep`**:
+
+```diff
+extension radiusCore
+
+@description('Information about what resource is calling this Recipe. Generated by Radius.')
+param context object
+
++resource gateway 'Applications.Core/gateways@2023-10-01-preview' = if (context.resource.properties.ingress) {
+  name: 'gateway'
+  properties: {
+    application: context.application.id
+    routes: [
+      {
+        path: '/'
+        destination: 'http://${serviceContainer.name}:${context.resource.properties.container.ports.http.containerPort}'
+      }
+    ]
+  }
+}
+
+resource serviceContainer 'Applications.Core/containers@2023-10-01-preview' = {
+  name: '${context.resource.name}-container'
+  properties: {
+    application: context.application.id
+    container: context.resource.properties.container
+    ...
+  }
+}
+```
+
+### User Story 7 – Registering recipes
+
+As a platform engineer, I need to register a recipe which implements my new resource type. 
+
+**Summary**
+
+Registering recipes works as expected just like existing portable resource types. However, since user-defined resource types introduces new capabilities, there are adjustments made to the recipe capabilities of Radius.
+
+**User Experience**
+
+```bash
+# Register the PostgreSQL Terraform recipe in the myEnv environment
+rad recipe register \
+  --environment myEnv \
+  --resource-type Radius.Resources/postgreSQL \
+  --recipe-kind terraform \
+  --recipe-path git://github.com/<username>/repo.git//recipes/postgreSQL
+Registering recipe for Radius.Resources/postgreSQL
+The recipe for Radius.Resources/postgreSQL is registered in the myEnv environment
+```
+
+There are two changes here:
+
+**Retirement of named recipes** – Today, Radius supports named recipes. When deploying a portable resource type, developers can override the default recipe by specifying a recipe name in their application definition. This enables developers to punch through the separation of concerns between platform and developers. Based on strong user feedback, user-defined resource types will not implement the ability to punch through and use alternate recipes. 
+
+Furthermore, once portable resource types are refactored into user-defined resource types, the ability to specify a recipe name will be retired. The command will change from ``rad recipe register <recipeName>` to `rad recipe register`. This also has the benefit of not having to build developer documentation for discovering available recipes. 
```

Comment:
I worry that taking away the ability to punch through will cause issues for users down the line who want to use this functionality in ways we don't fully understand yet